### PR TITLE
Add typing for User Management

### PR DESCRIPTION
**Pull request by mattgd**
## Description
Add typing for User Management.

## Documentation

Does this require changes to the WorkOS Docs? E.g. the [API Reference](https://workos.com/docs/reference) or code snippets need updates.

```
[ ] Yes
```

If yes, link a related docs PR and add a docs maintainer as a reviewer. Their approval is required.

### CODE REVIEW

File: workos/resources/user_management.py [hunk 1]
```diff
@@ -1,232 +1,122 @@
...
-
-    Attributes:
-        OBJECT_FIELDS (list): List of fields a WorkOSInvitation comprises.
-    """
+    access_token: str
+    refresh_token: str
+
+
+class EmailVerification(WorkOSModel):
+    """Representation of a EmailVerification object as returned by WorkOS through User Management features."""
```
Comment (PaulAsjes) [inline workos/resources/user_management.py:49]: Do we need to keep reminding developers that this is through User Management features?

File: workos/user_management.py [hunk 2]
```diff
@@ -1055,189 +981,128 @@ def create_password_reset(
...
+    def reset_password(self, token: str, new_password: str) -> User:
         """Resets user password using token that was sent to the user.
 
         Kwargs:
             token (str): The reset token emailed to the user.
             new_password (str): The new password to be set for the user.
 
         Returns:
-            dict: User response from WorkOS.
+            User: User response from WorkOS.
```
Comment (PaulAsjes) [inline workos/user_management.py:1006]: 👏

File: workos/user_management.py [hunk 3]
```diff
@@ -1312,120 +1144,118 @@ def enroll_auth_factor(
...
-            "params": {
-                "user_id": user_id,
-            },
-            "method": UserManagement.list_auth_factors,
+        # We don't spread params on this dict to make mypy happy
+        list_args: AuthenticationFactorsListFilters = {
+            "limit": limit or DEFAULT_LIST_RESPONSE_LIMIT,
+            "before": before,
+            "after": after,
+            "order": order or "desc",
```
Comment (PaulAsjes) [inline workos/user_management.py:1202]: Previously we changed this to `"order": order` dropping the `or "desc"` part. Should that be the case here?

**Review by PaulAsjes — CHANGES_REQUESTED** [review]
Hard to review a PR of this magnitude, but I found some nits that I think we should address.

File: workos/user_management.py [hunk 4]
```diff
@@ -58,223 +68,349 @@
...
+        last_name: Optional[str] = None,
+        email_verified: Optional[bool] = False,
+    ) -> User: ...
+
+    def update_user(
+        self,
+        user_id: str,
+        first_name: Optional[str] = None,
+        last_name: Optional[str] = None,
+        email_verified: Optional[bool] = False,
```
Comment (tribble) [inline workos/user_management.py:123]: Would this mean that by default we would always pass `False` on update? Would that end up unverifying a user?

File: workos/user_management.py [hunk 5]
```diff
@@ -58,223 +68,349 @@
...
+            params["provider"] = provider
+        if domain_hint is not None:
+            params["domain_hint"] = domain_hint
+        if login_hint is not None:
+            params["login_hint"] = login_hint
+        if state is not None:
+            params["state"] = state
+        if code_challenge:
+            params["code_challenge"] = code_challenge
+            params["code_challenge_method"] = "S256"
```
Comment (tribble) [inline workos/user_management.py:227]: Makes we wonder why we have a param for this if it only has one value

**Review by tribble — APPROVED** [review]
One thing I thing needs to be addressed: setting `email_verified=False` by default in user update methods. I'm worried that's going to lead to unexpected behavior.

Otherwise, I don't like that our API JSON models don't match our API reference docs. For now I'm OK going with whatever our ref docs say, but it seems like there should be a follow-up from the teams that own various API surface areas to eliminate the discrepancies.

Thoughts on ☝️ @PaulAsjes?

Slow smile

**Comment by PaulAsjes** [comment]
I agree that `email_verified` shouldn't default to anything. API ref says this field is optional, so we should send `undefined` if it's not explicitly set as presumably the API can handle that.

File: tests/conftest.py [hunk 6]
```diff
@@ -224,3 +223,31 @@ def mock_function(*args, **kwargs):
...
+
+@pytest.fixture
+def test_sync_auto_pagination(
+    mock_pagination_request_for_http_client,
+):
+    def inner(
+        http_client: Union[SyncHTTPClient, AsyncHTTPClient],
+        list_function: Callable[[], WorkOsListResource],
+        expected_all_page_data: dict,
+        list_function_params: Optional[Mapping[str, Any]] = {},
```
[the remark this answers is not in the document]
Comment (mattgd) [reply tests/conftest.py:236]: Yes we did, fixing.

File: tests/test_user_management.py [hunk 7]
```diff
@@ -652,261 +459,240 @@ def test_authorization_url_has_expected_query_params_with_code_challenge(self):
...
+        response = self.user_management.authenticate_with_code(**params)
 
-        assert request_args[1].endswith("user_management/authenticate")
-        assert response["user"]["id"] == "user_01H7ZGXFP5C6BBQY6Z7277ZCT0"
-        assert response["impersonator"]["email"] == "admin@foocorp.com"
-        assert response["impersonator"]["reason"] == "Debugging an account issue."
+        assert request_kwargs["url"].endswith("user_management/authenticate")
+        assert response.user.id == "user_01H7ZGXFP5C6BBQY6Z7277ZCT0"
+        assert response.impersonator is not None
+        assert response.impersonator.dict() == {
```
[the remark this answers is not in the document]
Comment (mattgd) [reply tests/test_user_management.py:536]: Hmm, yeah. A bit odd. These are sourced from the dashboard-only, so there's definitely some other magic going on.

File: workos/resources/mfa.py [hunk 8]
```diff
@@ -31,6 +57,33 @@ def to_dict(self):
...
+class AuthenticationFactorTotp(WorkOSModel):
+    """Representation of a MFA Authentication Factor Response as returned by WorkOS through the MFA feature."""
+
+    object: Literal["authentication_factor"]
+    id: str
+    created_at: str
+    updated_at: str
+    type: TotpAuthenticationFactorType
+    totp: Union[TotpFactor, ExtendedTotpFactor, None]
+    user_id: str
```
[the remark this answers is not in the document]
Comment (mattgd) [reply workos/resources/mfa.py:69]: Didn't see that, updating.

File: workos/resources/mfa.py [hunk 9]
```diff
@@ -91,6 +144,24 @@ def to_dict(self):
...
 
 
+class AuthenticationChallenge(WorkOSModel):
+    """Representation of a MFA Challenge Response as returned by WorkOS through the MFA feature."""
+
+    object: Literal["authentication_challenge"]
+    id: str
+    created_at: str
+    updated_at: str
+    expires_at: str
```
[the remark this answers is not in the document]
Comment (mattgd) [reply workos/resources/mfa.py:154]: Added/changed!

File: workos/resources/user_management.py [hunk 10]
```diff
@@ -1,232 +1,122 @@
...
-            authentication_response_dict["impersonator"] = self.impersonator.to_dict()
+class AuthenticationResponse(WorkOSModel):
+    """Representation of a User and Organization ID response as returned by WorkOS through User Management features."""
 
-        return authentication_response_dict
+    access_token: str
+    impersonator: Optional[Impersonator] = None
+    organization_id: Optional[str] = None
+    refresh_token: str
+    user: User
```
[the remark this answers is not in the document]
Comment (mattgd) [reply workos/resources/user_management.py:38]: Yes we do, added!

File: workos/resources/user_management.py (hunk 1 above)
Comment (mattgd) [reply workos/resources/user_management.py:49]: We don't. Removing that copy-pasta.

File: workos/resources/user_management.py [hunk 11]
```diff
@@ -1,232 +1,122 @@
...
+    updated_at: str
+
+
+class Invitation(WorkOSModel):
+    """Representation of an Invitation as returned by WorkOS through User Management features."""
+
+    object: Literal["invitation"]
+    id: str
+    email: str
+    state: str
```
[the remark this answers is not in the document]
Comment (mattgd) [reply workos/resources/user_management.py:67]: It should be. Fixed!

File: workos/user_management.py (hunk 4 above)
Comment (mattgd) [reply workos/user_management.py:123]: Yep, that's not correct. Should be `= None` as it's optional and we'll defer to API behavior.

File: workos/user_management.py [hunk 12]
```diff
@@ -283,121 +419,134 @@ def list_users(
...
-        return WorkOSUser.construct_from_response(response).to_dict()
+        return User.model_validate(response)
 
-    def update_user(self, user_id, payload):
+    def update_user(
+        self,
+        user_id: str,
+        first_name: Optional[str] = None,
+        last_name: Optional[str] = None,
+        email_verified: Optional[bool] = False,
```
[the remark this answers is not in the document]
Comment (mattgd) [reply workos/user_management.py:498]: Fixed!

File: workos/user_management.py (hunk 3 above)
Comment (mattgd) [reply workos/user_management.py:1209]: It should, fixed!

**Comment by tribble** [comment]
@PaulAsjes Longer term I'm more bothered by this:

> Otherwise, I don't like that our API JSON models don't match our API reference docs.

**Pull request merged**